src/nvim/eval.c

```diff
@@ -16753,6 +16855,83 @@ static void f_visualmode(typval_T *argvars, typval_T *rettv)
     curbuf->b_visual_mode_eval = NUL;
 }
 
+// watcheradd(dict, key, funcref) function
+static void f_watcheradd(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
+      (argvars[2].v_type != VAR_FUNC && argvars[2].v_type != VAR_STRING)) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
+    return;
+  }
+
+  if (!STRLEN(argvars[1].vval.v_string)) {
+    EMSG(_("Can't use an empty string as the watched key"));
+    return;
+  }
+
+  ufunc_T *func = find_ufunc(argvars[2].vval.v_string);
+  if (!func) {
+    // Invalid function name. Error already reported by `find_ufunc`.
+    return;
+  }
+
+  func->uf_refcount++;
+  DictWatcher *watcher = xmalloc(sizeof(DictWatcher));
+  watcher->key_pattern = xstrdup((char *)argvars[1].vval.v_string);
+  watcher->callback = func;
+  watcher->busy = false;
+  QUEUE_INSERT_TAIL(&argvars[0].vval.v_dict->watchers, &watcher->node);
+}
+
+// watcherdel(dict, key, funcref) function
+static void f_watcherdel(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
+      (argvars[2].v_type != VAR_FUNC && argvars[2].v_type != VAR_STRING)) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
+    return;
+  }
+
+  ufunc_T *func = find_ufunc(argvars[2].vval.v_string);
+  if (!func) {
+    // Invalid function name. Error already reported by `find_ufunc`.
+    return;
+  }
+
+  dict_T *dict = argvars[0].vval.v_dict;
+  QUEUE *w = NULL;
+  DictWatcher *watcher = NULL;
+  bool matched = false;
+  QUEUE_FOREACH(w, &dict->watchers) {
```

Comment:
I predict Neovim-specific optimization principle: “don’t touch global and buffer vars, if you must squash all settings in one `g:` subdictionary”. If this mechanizm is expected to be used by plugins for their setting then there would be *lots* of watchers over time watching keys in `g:`. Hash servers much better here then a queue with `O(n)` querying.